waku/v2/peermanager/rtt_cache.go

```diff
@@ -0,0 +1,256 @@
+	if len(r.pingQueue) == 0 {
+		toPing = r.host.Peerstore().Peers()
+	} else {
+		for _, p := range r.pingQueue {
+			if p.IsStale() {
+				toPing = append(toPing, p.PeerID)
+			}
+		}
+	}
+	r.RUnlock()
+	for _, p := range toPing {
+		go r.PingPeer(ctx, p)
+	}
+}
+
+func (r *RTTCache) start(ctx context.Context, pingInterval time.Duration) {
+	ticker := time.NewTicker(pingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			r.pingPeers(ctx)
+		}
+	}
+
+}
+
+func (r *RTTCache) PingPeer(ctx context.Context, peer peer.ID) (*PeerRTT, error) {
+	if peer == r.host.ID() {
+		return nil, nil // Don't ping yourself
+	}
+
+	r.RLock()
+	peerRTT, ok := r.peers[peer]
+	r.RUnlock()
+
+	if ok && !peerRTT.IsStale() {
+		// We already have a ping recorded
+		return peerRTT, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+
+	case result := <-ping.Ping(ctx, r.host, peer):
+		r.Lock()
+		defer r.Unlock()
+
+		peerRTT, exists := r.getOrDefault(peer)
+		if result.Error == nil {
+			peerRTT.PingRTT = result.RTT
+			peerRTT.nextVerification = time.Now().Add(r.verificationInterval)
+			if exists {
+				r.pingQueue.Update(peerRTT)
+			} else {
+				r.pingQueue.Push(peerRTT)
+			}
+			return peerRTT, nil
+		} else {
+			peerRTT.PingRTT = time.Hour // This is just so we don't choose this peer again since the RTT is unreasonable
+			peerRTT.nextVerification = time.Now().Add(peerRTT.bkf.Delay())
+			if exists {
+				r.pingQueue.Update(peerRTT)
+			} else {
+				r.pingQueue.Push(peerRTT)
+			}
+
+			r.logger.Debug("could not ping", logging.HostID("peer", peer), zap.Error(result.Error))
+
+			return nil, result.Error
+		}
+	}
+}
+
+func (r *RTTCache) PeersRTT(ctx context.Context, peers peer.IDSlice) []*PeerRTT {
+	wg := sync.WaitGroup{}
+	pingCh := make(chan *PeerRTT, len(peers))
+	wg.Add(len(peers))
+	for _, p := range peers {
+		go func(p peer.ID) {
+			defer wg.Done()
+			peerRTT, err := r.PingPeer(ctx, p)
+			if err == nil {
+				pingCh <- peerRTT
+			}
+		}(p)
+	}
+
+	wg.Wait()
+	close(pingCh)
+
+	var result []*PeerRTT
+	for peerRTT := range pingCh {
+		result = append(result, peerRTT)
+	}
+
+	return result
+}
+
+func (r *RTTCache) FastestPeer(ctx context.Context, peers peer.IDSlice) (*PeerRTT, error) {
+	var min *PeerRTT
+	for _, p := range r.PeersRTT(ctx, peers) {
+		if min == nil {
+			min = p
+		} else {
+			if p.PingRTT < min.PingRTT {
+				min = p
+			}
+		}
+	}
+	if min == nil || min.PingRTT == time.Hour {
+		return nil, ErrNoPeersAvailable
+	}
+
+	return min, nil
+}
+
+func (r *RTTCache) getOrDefault(peerID peer.ID) (*PeerRTT, bool) {
+	peerRTT, ok := r.peers[peerID]
+	exists := true
+
+	if !ok {
+
+		rngSrc := rand.NewSource(rand.Int63())
+		minBackoff, maxBackoff := time.Second*5, time.Hour
+		bkf := backoff.NewExponentialBackoff(minBackoff, maxBackoff, backoff.FullJitter, time.Second, 5.0, 0, rand.New(rngSrc))()
```

Comment:
Maybe the backoff growth could be based on verificationInterval rather than hrdcoded to 5secs?